### PR TITLE
Ips handling/v24

### PR DESCRIPTION
**Pull request by victorjulien**
Implements a new way of dealing with memcap hits: a per subsystem `memcap-policy` which can be `drop-packet`, `drop-flow`, `pass-packet`, `pass-flow`, `bypass` or `ignore`. Where `ignore` is essentially the current behavior. Looking for a better name and general feedback on the approach.

In https://github.com/OISF/suricata-verify/pull/818 there are tests for most options in combination with stream reassembly memcap failure simulation.

continues work in #7234 
(includes #7356)

suricata-verify-pr: 818

### CODE REVIEW

File: src/util-memcap.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    const char *value_str = NULL;
+    if ((ConfGetValue(option, &value_str)) == 1 && value_str != NULL) {
+        if (strcmp(value_str, "drop-flow") == 0) {
+            policy = MEMCAP_POLICY_DROP_FLOW;
+            SCLogNotice("%s: %s", option, value_str);
+        } else if (strcmp(value_str, "pass-flow") == 0) {
+            policy = MEMCAP_POLICY_PASS_FLOW;
+            SCLogNotice("%s: %s", option, value_str);
+        } else if ((strcmp(value_str, "bypass") == 0) ||
+                   (strcmp(value_str, "bypass-flow") == 0)) {
```
Comment (jasonish) [inline src/util-memcap.c:76]: Both names required? Or can we just use one?

File: src/stream-tcp.c [hunk 2]
```diff
@@ -462,6 +463,9 @@ void StreamTcpInitConfig(bool quiet)
             stream_config.flags |= STREAMTCP_INIT_FLAG_INLINE;
         }
     }
+    stream_config.ssn_memcap_policy = MemcapPolicyParse("stream.memcap-policy");
+    stream_config.reassembly_memcap_policy = MemcapPolicyParse("stream.reassembly.memcap-policy");
```
Comment (jasonish) [inline src/stream-tcp.c:467]: Or should policies be in their own configuration file section:
```
policy:
  stream-memcap: ...
  stream-reassembly-memcap: ...
```

**Comment by jasonish** [comment]
> On further thought I think the MemcapPolicy thing could even be a more general ExceptionPolicy, where possible exceptions include memcap hits, other resource limit hits, app errors and probably more

Yes, I agree.  I also think it would be more user friendly if this exception policy was configured in one place, rather than having it sprinkled among the various modules.

**Comment by victorjulien** [comment]
> > On further thought I think the MemcapPolicy thing could even be a more general ExceptionPolicy, where possible exceptions include memcap hits, other resource limit hits, app errors and probably more
> 
> Yes, I agree. I also think it would be more user friendly if this exception policy was configured in one place, rather than having it sprinkled among the various modules.

Tough one. I feel like it makes sense to have them close to the memcap settings, but I also see the point of grouping them. Ultimately I feel most ppl would never touch these. They would get the implied values (`ignore` for IDS, `drop-flow` for IPS). The actual settings would be available for expert tuning.

File: src/util-memcap.c (hunk 1 above)
Comment (victorjulien) [reply src/util-memcap.c:76]: Can do one. We use `bypass` in various places in config, but here I used `flow-(packet|flow)` pattern, so it seemed to make sense to have both.

Comment (jasonish) [reply src/util-memcap.c:76]: Being a new configuration option, I'd rather not see an alias right from the start.

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpw1_files_sha256.
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpr1_stats_chk | | |
| .app_layer.error.ftp-data.parser | 0 | 395 | 0.0% | 

Pipeline 7353

File: src/util-memcap.c (hunk 1 above)
Comment (victorjulien) [reply src/util-memcap.c:76]: went with just `bypass`

**Pull request closed without merging**